gcp : add alloydb group

jcockbain opened the pull request:
<!--
Thank you for helping to improve Official GCP Provider!

Please read through https://git.io/fj2m9 if this is your first time opening a
Official GCP Provider pull request. Find us in https://slack.crossplane.io/messages/dev if
you need any help contributing.
-->

### Description of your changes

<!--
Briefly describe what this pull request does. Be sure to direct your reviewers'
attention to anything that needs special consideration.

We love pull requests that resolve an open Official GCP Provider issue. If yours does, you
can uncomment the below line to indicate which issue your PR fixes, for example
"Fixes #500":

-->
Fixes #371

Adds resources for: 

[google_alloydb_backup](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/alloydb_backup)
[google_alloydb_instance](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/alloydb_cluster)
[google_alloydb_cluster](https://registry.terraform.io/providers/hashicorp/google/latest/docs/resources/alloydb_cluster)

I have:

- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

<!--
Before reviewers can be confident in the correctness of this pull request, it
needs to tested and shown to be correct. Briefly describe the testing that has
already been done or which is planned for this change.
-->

I deployed instances of the 3 new kinds to to a minikube cluster. These were based on the generated examples, adapted for my test GCP project. 

I then used `make run`, which created the expected resources in my project.

Each of these resource became "ready". Deleting the k8s resources deletes the resources from GCP.  

Cluster: https://github.com/upbound/provider-gcp/actions/runs/6110391551
![Screenshot from 2023-09-06 17-48-49](https://github.com/upbound/provider-gcp/assets/31372187/c6af8d86-0df9-42d3-a8cd-b5e486cd51eb)

Instance: 
![Screenshot from 2023-09-07 11-44-35](https://github.com/upbound/provider-gcp/assets/31372187/b44fbde2-1e6c-4b5f-bf2b-a3c37b112ba7)

Backup:
![Screenshot from 2023-09-07 13-46-57](https://github.com/upbound/provider-gcp/assets/31372187/68bdd2f9-fdc8-474a-b908-c459b2d2e42f)

Upbound-CLA commented:
[![CLA assistant check](https://cla.upbound.io/pull/badge/not_signed)](https://cla.upbound.io/upbound/provider-gcp?pullRequest=373) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.upbound.io/upbound/provider-gcp?pullRequest=373) before we can accept your contribution.<br/><hr/>**jcockbain** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.upbound.io/check/upbound/provider-gcp?pullRequest=373) it.</sub>

turkenf commented:
/test-examples="examples/alloydb/instance.yaml"

jcockbain commented:
@turkenf are you able to assist with the failure on `ci/check-diff`? Running `make check-diff` from my workstation is passes fine.

turkenf commented:
Hi @jcockbain, 

After running make generate there seem to be changes you didn't commit, can you run make generate again and commit all the changes?


![Screenshot 2023-09-04 at 16 11 09](https://github.com/upbound/provider-gcp/assets/103541666/b25a03fb-2dfb-4402-814a-07f6e82dca7a)

jcockbain commented:
@turkenf thanks. I have updated my branch, and changed the commits to show the manual and then the generated changes. `make generate` does not create any new files on my branch now, so hopefully the CI will be okay.

turkenf commented:
@jcockbain, could you please rebase your branch to the main, run make generate, and commit all the changes again?

turkenf commented on config/externalname.go:
```diff
@@ -14,6 +14,15 @@ var externalNameConfigs = map[string]config.ExternalName{
+	// Imported by using the following format: projects/{{project}}/locations/{{location}}/clusters/{{cluster_id}}
+	"google_alloydb_cluster": config.TemplatedStringAsIdentifier("cluster_id", "projects/{{ .setup.configuration.project }}/location/{{ .parameters.location }}/clusters/{{ .external_name }}"),
+	// Imported by using the following format: {{cluster}}//{{instance_id}}
+	"google_alloydb_instance": config.IdentifierFromProvider,
```
```suggestion
	"google_alloydb_instance": config.TemplatedStringAsIdentifier("instance_id", "{{ .parameters.cluster }}/{{ .external_name }}"),
```

Could you please try this configuration, if it works it could be better.

turkenf commented on config/externalname.go:
```diff
@@ -14,6 +14,15 @@ var externalNameConfigs = map[string]config.ExternalName{
+	// Imported by using the following format: {{cluster}}//{{instance_id}}
+	"google_alloydb_instance": config.IdentifierFromProvider,
+	// Imported by using the following format: projects/{{project}}/locations/{{location}}/backups/{{backup_id}}
+	"google_alloydb_backup": config.TemplatedStringAsIdentifier("backup_id", "projects/{{ .setup.configuration.project }}/location/{{ .parameters.location }}/clusters/{{ .external_name }}"),
```
```suggestion
	"google_alloydb_backup": config.TemplatedStringAsIdentifier("backup_id", "projects/{{ .setup.configuration.project }}/location/{{ .parameters.location }}/backups/{{ .external_name }}"),
```

turkenf commented on examples/alloydb/cluster.yaml:
```diff
@@ -0,0 +1,25 @@
+spec:
+  forProvider:
+    location: us-central1
+    network: projects/${data.google_project.project.number}/global/networks/${google_compute_network.default.name}
```
```suggestion
    networkSelector:
      matchLabels:
        testing.upbound.io/example-name: default
```

It is better to use a selector so that our examples can be tested automatically via uptest.

turkenf commented on examples/alloydb/backup.yaml:
```diff
@@ -0,0 +1,85 @@
+spec:
+  forProvider:
+    location: us-central1
+    network: ${data.google_compute_network.default.id}
```
```suggestion
    networkSelector:
      matchLabels:
        testing.upbound.io/example-name: google_compute_network
```

Same here and we should add `Network.compute` resource in this example manifest.

turkenf commented on examples/alloydb/instance.yaml:
```diff
@@ -0,0 +1,88 @@
+        name: example-secret
+        namespace: upbound-system
+    location: us-central1
+    network: projects/${data.google_project.project.number}/global/networks/${google_compute_network.default.name}
```
```suggestion
    networkSelector:
      matchLabels:
        testing.upbound.io/example-name: default
```

turkenf reviewed (commented):
@jcockbain, many thanks for your contribution. I left a few comments for you to consider. 

Could you please fill in how it was tested after the last changes for each resource as [here](https://github.com/upbound/provider-gcp/pull/229#issue-1576825441)?

jcockbain commented:
Thanks very much for your comments @turkenf, I have attempted to resolve them all in the above commit. 

I also retested with the examples and added screenshots to the summary of this PR.

jcockbain commented:
One thing I am not certain on is what to name example resources, for my testing I gave the resources different names as `default` is already used for some of the resources in my test project. So I left these test names in the example manifests, but I can change them back to the generated ones if that would be better.

turkenf commented on examples/alloydb/backup.yaml:
```diff
@@ -0,0 +1,99 @@
+kind: Network
+metadata:
+  annotations:
+    meta.upbound.io/example-id: alloydb/v1beta1/cluster
```
```suggestion
    meta.upbound.io/example-id: alloydb/v1beta1/backup
```

turkenf commented on examples/alloydb/instance.yaml:
```diff
@@ -0,0 +1,89 @@
+kind: Network
+metadata:
+  annotations:
+    meta.upbound.io/example-id: alloydb/v1beta1/cluster
```
```suggestion
    meta.upbound.io/example-id: alloydb/v1beta1/instance
```

turkenf reviewed (commented):
Thank you for the quick update @jcockbain, the names look good for now, they can stay that way. I just left two minor comments that fix the example-ids.

turkenf commented:
/test-examples="examples/alloydb/cluster.yaml"

turkenf commented:
/test-examples="examples/alloydb/cluster.yaml"

turkenf reviewed (approved):
@jcockbain, Thank you again for your contribution and clean work. LGTM.
